launcher/internal/gpu/driverinstaller.go

```diff
@@ -0,0 +1,122 @@
+func NewDriverInstaller(cdClient *containerd.Client, launchSpec spec.LaunchSpec, logger *log.Logger) *DriverInstaller {
+	return &DriverInstaller{
+		cdClient:   cdClient,
+		launchSpec: launchSpec,
+		logger:     logger,
+	}
+}
+
+// InstallGPUDrivers installs the GPU driver on host machine using the cos-gpu-installer container.
+// It uses containerd package to launch and manage the driver installer container.
+func (di *DriverInstaller) InstallGPUDrivers(ctx context.Context) error {
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
+	installerImageRef, err := getInstallerImageReference()
+	if err != nil {
+		di.logger.Printf("failed to get the installer container image reference: %v", err)
+		return err
+	}
+
+	di.logger.Printf("cos gpu installer version : %s", installerImageRef)
+	image, err := di.cdClient.Pull(ctx, installerImageRef, containerd.WithPullUnpack)
+	if err != nil {
+		return fmt.Errorf("failed to pull installer image: %v", err)
+	}
+
+	mounts := []specs.Mount{
+		{
+			Type:        "volume",
+			Source:      "/dev",
+			Destination: "/dev",
+			Options:     []string{"rbind", "rw"},
+		}, {
+			Type:        "volume",
+			Source:      "/",
+			Destination: "/root",
+			Options:     []string{"rbind", "rw"},
+		},
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		di.logger.Printf("cannot get hostname: %v", err)
+	}
+
+	container, err := di.cdClient.NewContainer(
+		ctx,
+		installerContainerID,
+		containerd.WithImage(image),
+		containerd.WithNewSnapshot(installerSnapshotID, image),
+		containerd.WithNewSpec(oci.WithImageConfig(image),
+			oci.WithPrivileged,
+			oci.WithProcessArgs("/cos-gpu-installer", "install", "-version=default", fmt.Sprintf("-host-dir=%s", InstallationHostDir)),
+			oci.WithAllDevicesAllowed,
+			oci.WithMounts(mounts),
+			oci.WithHostNamespace(specs.NetworkNamespace),
+			oci.WithHostNamespace(specs.PIDNamespace),
+			oci.WithHostHostsFile,
+			oci.WithHostResolvconf,
+			oci.WithEnv([]string{fmt.Sprintf("HOSTNAME=%s", hostname)})))
+	if err != nil {
+		return fmt.Errorf("failed to create gpu driver installer container: %v", err)
+	}
+
+	task, err := container.NewTask(ctx, cio.NewCreator(cio.WithStdio))
+	if err != nil {
+		return fmt.Errorf("failed to create gpu driver installation task: %v", err)
+	}
+
+	statusC, err := task.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to wait for gpu driver installation task: %v", err)
+	}
+
+	if err := task.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start gpu driver installation task: %v", err)
+	}
+
+	status := <-statusC
+	code, _, _ := status.Result()
+	di.logger.Printf("Gpu driver installation task exited with status: %d\n", code)
+	return nil
+}
+
+func getInstallerImageReference() (string, error) {
+	installerImageRefBytes, err := exec.Command("cos-extensions", "list", "--", "--gpu-installer").Output()
+	installerImageRef := strings.TrimSpace(string(installerImageRefBytes))
+	if err != nil {
```

Comment:
move the error check up, also doesn't make sense to have installerImageRef in the error message

Reply:
done